common/collection/concurrent_tx_map.go

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package collection
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+const (
+	// nShards represents the number of shards
+	// At any given point of time, there can only
+	// be nShards number of concurrent writers to
+	// the map at max
+	nShards = 32
+)
+
+type (
+	// HashFunc represents a hash function for string
+	HashFunc func(interface{}) uint32
+
+	// ConcurrentTxMap is a generic interface for any
+	// implementation of a dictionary or a key value
+	// lookup table that is thread safe, and
+	ConcurrentTxMap interface {
+		// Get returns the value for the given key
+		Get(key interface{}) (interface{}, bool)
+		// Contains returns true if the key exist and false otherwise
+		Contains(key interface{}) bool
+		// Put records the mapping from given key to value
+		Put(key interface{}, value interface{})
+		// PutIfNotExist records the key value mapping only
+		// if the mapping does not already exist
+		PutIfNotExist(key interface{}, value interface{}) bool
+		// Remove deletes the key from the map
+		Remove(key interface{})
+		// GetAndDo returns the value corresponding to the key, and apply fn to key value before return value
+		GetAndDo(key interface{}, fn func(key interface{}, value interface{})) (interface{}, bool)
```

Comment:
Can you define a type for Action and Predicate and use it here.

Reply:
done